### PR TITLE
Add skew-symmetric DG implementation

### PR DESCRIPTION
**Pull request by peterrum**
This PR adds a skew-symmetric DG implementation to the advection operation. This involves, the addition of following terms:

<p align="center">
 
</p>

with the skew factor .

### CODE REVIEW

File: include/hyper.deal/matrix_free/fe_evaluation_cell.h [hunk 1]
```diff
@@ -197,6 +197,88 @@ namespace hyperdeal
...
+    inline DEAL_II_ALWAYS_INLINE //
+      dealii::Tensor<1, dim_v, VectorizedArrayType>
+      get_gradient_v(const VectorizedArrayType *__restrict grad_in,
+                     const unsigned int q,
+                     const unsigned int qx,
+                     const unsigned int qv) const
+    {
+      (void)qx;
+
+      dealii::Tensor<1, dim_x, VectorizedArrayType> result;
```
Comment (kronbichler) [inline include/hyper.deal/matrix_free/fe_evaluation_cell.h:241]:
```suggestion
      dealii::Tensor<1, dim_v, VectorizedArrayType> result;
```

File: include/hyper.deal/matrix_free/fe_evaluation_cell.h [hunk 2]
```diff
@@ -197,6 +197,88 @@ namespace hyperdeal
...
+                          grad_in[q + e * n_q_points_x * n_q_points_v]);
+        }
+
+      return result;
+    }
+
+
+
+    /**
+     * Submit value in x-space
```
Comment (kronbichler) [inline include/hyper.deal/matrix_free/fe_evaluation_cell.h:260]:
```suggestion
     * Submit value for integration along x-v coordinates
```

File: include/hyper.deal/operators/advection/advection_operation.h [hunk 3]
```diff
@@ -303,31 +308,61 @@ namespace hyperdeal
             dealii::AlignedVector<VNumber> scratch_data_array;
             scratch_data_array.resize_fast(dealii::Utilities::pow(n_points, dim) * dim_x);
             VNumber *tempp = scratch_data_array.begin();
+            
+            if(do_skew)
```
Comment (kronbichler) [inline include/hyper.deal/operators/advection/advection_operation.h:312]: Why not skip the variable `do_skew` and simply check
```suggestion
            if(factor_skew != 0.0)
```
(I would also check for `!=` because in principle one could imagine negative factors even though I'm not sure it makes sense.)

File: include/hyper.deal/operators/advection/advection_operation.h [hunk 4]
```diff
@@ -303,31 +308,61 @@ namespace hyperdeal
...
             for (auto qv = 0u, q = 0u; qv < dealii::Utilities::pow<unsigned int>(n_points, dim_v); ++qv)
               for (auto qx = 0u; qx < dealii::Utilities::pow<unsigned int>(n_points, dim_x); ++qx, ++q)
                 {
                   VNumber    grad_in[dim_x];
                   const auto vel = velocity_field->evaluate_x(q, qx, qv);
+                  
+                  if(do_skew)
+                    phi.template submit_value<false>(data_ptr, - factor_skew * (phi.get_gradient_x(tempp, q, qx, qv) * vel), q, qx, qv);
+                  
                   for (int d = 0; d < dim_x; d++)
```
Comment (kronbichler) [inline include/hyper.deal/operators/advection/advection_operation.h:328]:
```suggestion
                  if (factor_skew != 1.0)
                    for (int d = 0; d < dim_x; d++)
```

File: include/hyper.deal/operators/advection/advection_operation.h [hunk 5]
```diff
@@ -303,31 +308,61 @@ namespace hyperdeal
...
-                    grad_in[d] = buffer[q] * vel[d];
+                    grad_in[d] = (1.0-factor_skew) * buffer[q] * vel[d];
                   phi.submit_gradient_x(tempp, grad_in, q, qx, qv);
                 }
-            if (dim_x >= 1) eval_.template gradients<0, false, false>(tempp + dealii::Utilities::pow(n_points, dim) * 0, data_ptr);
+
```
Comment (kronbichler) [inline include/hyper.deal/operators/advection/advection_operation.h:332]: The part starting from here and up to line 339 should be incapsulated into `if (factor_skew != 1.0)`

File: include/hyper.deal/operators/advection/advection_operation.h [hunk 6]
```diff
@@ -303,31 +308,61 @@ namespace hyperdeal
...
             if (dim_x >= 3) eval_.template gradients<2, false, true >(tempp + dealii::Utilities::pow(n_points, dim) * 2, data_ptr);
           }
+            
           // v-space
           {
             dealii::AlignedVector<VNumber> scratch_data_array;
             scratch_data_array.resize_fast(dealii::Utilities::pow(n_points, dim) * dim_v);
             VNumber *tempp = scratch_data_array.begin();
+            
+            if(do_skew)
```
Comment (kronbichler) [inline include/hyper.deal/operators/advection/advection_operation.h:348]: Same as above.

**Review by kronbichler — APPROVED** [review]
Looks good apart from a few comments.

**Comment by peterrum** [comment]
Thanks @kronbichler for the review! I have added the suggestions. I will leave the implementation of boundary conditions and FCL for a follow-up PR. I'll merge once the tests are happy.

File: examples/advection/tests/adv_1D_1D_k3.hyperrectangle_03.json [hunk 7]
```diff
@@ -32,6 +32,9 @@
         "PerformanceLogAllCalls": "false",
         "DiagnosticsWarmUpIterations": "0"
     },
+    "AdvectionOperation": {
+        "SkewFactor": "0.0"
+    },
```
Comment (peterrum) [inline examples/advection/tests/adv_1D_1D_k3.hyperrectangle_03.json:37]: I had to add these lines. Because else I get segmentation fault on Travis CI!? This is definitely a bug. However, let's merge and tackle this issue later!
